sycl/source/detail/program_impl.cpp

```diff
@@ -162,6 +165,7 @@ cl_program program_impl::get() const {
 void program_impl::compile_with_kernel_name(string_class KernelName,
                                             string_class CompileOptions,
                                             OSModuleHandle M) {
+  std::unique_lock<std::mutex> Lock(MMutex);
```

Comment:
It's much better to use std::lock_guard where you won't unlock the lock until end of scope